add setting [printer].mainsync to select mcu for main sync

hg42 opened the pull request:
This allows to select the mcu section that provides the clock synchronization master.

I also tried code to automatically default to the mcu handling stepper_x by extracting the mcu name from step_pin, but I think this solution is much shorter, simpler, more clear and has less dependencies.

This also removes the need for an anonymous mcu section.

Usage:
```
[printer]
mainsync = mcu ramps
```

Default is "mcu".

Signed-off-by: Harald Gutsche <hg42@gmx.net>

KevinOConnor commented:
Thanks.  However, this patch is not correct.  There are places outside of the mcu code that rely on the mainsync being attributed to "mcu".  In particular, the toolhead class does all of its time synchronization to the first mcu object returned by self.printer.lookup_module_objects('mcu') - it would not operate correctly if the first object returned is not the main "mcu".

Separately, what is the gain to implementing this?  In general, I prefer not to add code that allows for arbitrarily different ways to define the same thing.  As that tends to make handling of support requests harder (for example, when tracking down a bug, the person doing the error analysis has to consider the possibility that slightly different config choices could lead to slightly different behavior).

-Kevin

hg42 commented:
ok, I searched, but didn't find this.
May be the [mcu xxx] could also be stored as [mcu].

The main benefit is to change the synchronization master (e.g. if stepper_x/y is assigned to another mcu) without renaming both mcus and all their pins. Also renaming xxx:* to yyy:* is much easier than renaming from un-prefixed pin names.

hg42 commented:
> In particular, the toolhead class does all of its time synchronization to the first mcu object returned by self.printer.lookup_module_objects('mcu')

The code in toolhead gives a hint that you didn't assume a specific name for the mainsync mcu. Instead you used the mcu that was first added to `printer`.
So, I think, it should be safe to add the master mcu first like you did with the anonymous mcu (and excluding it in the loop).
I already have code and will probably test it tomorrow.

I did not find any other part relying on [mcu], neither via `lookup_module_objects` nor `lookup_object`.
Can I assume you always used the same access method for the same object?

Otherwise, could you give a hint where or how to search for other occurences?

hg42 commented:
[2018-06-26-151501-klippy.log](https://github.com/KevinOConnor/klipper/files/2137265/2018-06-26-151501-klippy.log)

this seems to work...

The behavior wasn't different from before, so the test may not be sufficient.
However, ToolHead only uses lines like
```
        if self.mcu.is_fileoutput():
```
to check for output going to a file instead of to the serial port for debugging and lines like
```
       est_print_time = self.mcu.estimated_print_time(self.reactor.monotonic())
```
How I interpret the usage of estimated print time, it doesn't seem to make much of a difference if the clocks are in sync.
![2018-06-26-151501-klippy-usage](https://user-images.githubusercontent.com/4707597/41916236-e72b481c-7957-11e8-8625-3e869076c837.png)
![2018-06-26-151501-klippy-frequency](https://user-images.githubusercontent.com/4707597/41916241-e940625e-7957-11e8-99e5-c7f4a15e3010.png)

KevinOConnor commented:
On Mon, Jun 25, 2018 at 12:01:53PM -0700, Harald wrote:
> The main benefit is to change the synchronization master (e.g. if stepper_x/y is assigned to another mcu) without renaming both mcus and all their pins. Also renaming xxx:* to yyy:* is much easier than renaming from un-prefixed pin names.

Okay.  That would be convenient.

However, the documentation would need to be updated (Code_Overview.md,
and example.cfg would need to be worded with a description that a user
not already familiar with Klipper could get a basic understanding
from).  I understand that not everyone is a native English speaker, so
I don't mind writing the documentation.  However, this isn't a
priority for me right now.  I'll try to look at it in a couple of
weeks.


On Mon, Jun 25, 2018 at 06:59:40PM -0700, Harald wrote:
> I did not find any other part relying on [mcu], neither via `lookup_module_objects` nor `lookup_object`.
> Can I assume you always used the same access method for the same object?

I'm not aware of any other uses.  It would require a bunch of testing.


On Tue, Jun 26, 2018 at 06:45:28AM -0700, Harald wrote:
> The behavior wasn't different from before, so the test may not be sufficient.

I suspect it could result in the odd buffer_time behavior you raised
in the smoothieboard issue.  That has me very concerned.

-Kevin

hg42 commented:
the whole issue could be simplified by keeping the slope limit (selecting the local maximum or minimum) and only providing access to the delta limit. While this doesn't select the shortest possible startup time, it is not much of a difference.
The delta is easier to describe and should be understandable without math knowledge.

This would also prevent starting the print in cases when the overshoot would exceed max_temp and shutdown Klipper.

The slope could still be made accessible for special purposes, but need not be documented for end users.

hg42 commented:
sorry I posted to the wrong issue...

hg42 commented:
It looks like I mixed up issues again...ignore my last comment(s).

> I suspect it could result in the odd buffer_time behavior you raised
in the smoothieboard issue

for this test I indeed used your branch together with only the (former) mainsync patch.
However, I used only a single mcu in all the smoothieboard tests, so syncing shouldn't matter at all, right?

hg42 commented:
when I changed to your branch, I changed my build system to work with an anonymous mcu section (not as difficult as I thought).
I also got some motors so I could restore my smoothieboard test environment.

I first made a test with the source code from https://github.com/KevinOConnor/klipper/issues/60#issuecomment-392387900 :+1:
[2018-06-28-132918-klippy--namedmcu-gcc493-mainsync-patched.log](https://github.com/KevinOConnor/klipper/files/2145231/2018-06-28-132918-klippy--namedmcu-gcc493-mainsync-patched.log)

Then I used the parent version of the patch:
[2018-06-28-135329-klippy--anonmcu-gcc493-c9dfdf21.log](https://github.com/KevinOConnor/klipper/files/2145240/2018-06-28-135329-klippy--anonmcu-gcc493-c9dfdf21.log)

Then I updated to your current master branch and also updated to gcc 6.3.1:

[2018-06-28-141538-klippy--anonmcu-gcc631-f529d747.log](https://github.com/KevinOConnor/klipper/files/2145249/2018-06-28-141538-klippy--anonmcu-gcc631-f529d747.log)

I also tested gcc 4.9.3 before with the same behavior, but the log was deleted.

All are tested with the same parameters. All show the same behavior.
The peaks seem to be at the same places (visually), if you remove the part up to the first peak (these part is different because of some manual actions, setting speed factor etc.).

Summary: the buffer_time issue also exists without the mainsync patch.

hg42 commented:
instead of `mainsync = section name` we could use a `mainsync = True` in the mcu section.
This would be easier for the user, because it removes the problem to exactly write the mcu section name (count of spaces).
However the code becomes slightly more complex.

hg42 commented:
could we change mainsync to something different? I think e.g. sync_master would be more clear?

hg42 commented:
it's the same like in #60 
( https://github.com/KevinOConnor/klipper/files/2042795/TX_bar-80x10x10-4--fs0.5--only-XYZ-no-homing-no-e.gcode.zip ), 
with only one change, I removed the slower first layer to save some time (note, the file only moves the head, no homing, no heaters, no extrusion).

If you want the real file, I can upload it, when I'm at home again.

Also note, I am not really complaining about the behavior, because the speeds are not realistic.
The main purpose of these tests are to verify that the issue is not caused by the mainsync patch.

Though, I think, the step rate could be higher (see some former comments #60).

hg42 commented:
here it is:

[TX_bar-80x10x10-4--fs0.5--only-XYZ-no-homing-no-e--2.gcode.zip](https://github.com/KevinOConnor/klipper/files/2152219/TX_bar-80x10x10-4--fs0.5--only-XYZ-no-homing-no-e--2.gcode.zip)

hg42 commented:
btw. in my tests I start the print and then issue `M220 S10000` to run at maximum speed (limited by the config file)

KevinOConnor commented:
On Sat, Jun 30, 2018 at 11:06:37AM -0700, Harald wrote:
> it's the same like in #60 
> ( https://github.com/KevinOConnor/klipper/files/2042795/TX_bar-80x10x10-4--fs0.5--only-XYZ-no-homing-no-e.gcode.zip ), 
> with only one change, I removed the slower first layer to save some time (note, the file only moves the head, no homing, no heaters, no extrusion).
> 
> If you want the real file, I can upload it, when I'm at home again.
> 
> Also note, I am not really complaining about the behavior, because the speeds are not realistic.
> The main purpose of these tests are to verify that the issue is not caused by the mainsync patch.

It looks like the jumps in print_time, buffer_time, and stalled_bytes
are due to the details of the test case.  Specifically, after any full
flush of the look-ahead queue, the queue will attempt to prime itself
with at least 2 seconds of moves before generating steps.  A simple
heuristic is used to gather at least 2 seconds of moves - the minimum
time of a move is calculated as move_distance/requested_velocity, and
moves are collected until the minimum move times sum to 2 seconds.

With an unrealistic requested move speed of 10,000mm/s, however, that
heuristic significantly underestimates the move time, and as a result
the look-ahead queue batches up a lot of moves before resuming step
generation.  When those steps are then generated, they result in a
jump in print_time, buffer_time, and stalled_bytes.  There's nothing
really wrong with that though.

-Kevin

hg42 commented:
ok, I think I understand...the 10000 mm/s is never reached, but it's used for planning and this fills the buffer.
So there is no need to change anything.

To me, one effect remains unexplained...
Looking at https://github.com/KevinOConnor/klipper/issues/60#issuecomment-392387900, with accel=10000
* a speed of 5000 creates continuous buffer_time peaks ("saw teeth") and a pause before each peak (where the buffer seems to be loaded)
* with a speed of 4000 there are still peaks (a little bit lower), but there is no pause before a peak. Also, buffer_time never goes down to 0.000 (which is coincident with pausing).

I don't finally understand, why this is happening and why there is a kind of limit (related to the height of the peaks) above which the buffer seems to be filled but printing pauses, while below the limit the buffer is also filled but printing continues. Eventually this could give a hint why the pausing happens at all.
I think, usually the buffer should restart filling during the decay phase (when it falls below 2 seconds?).
But when operating above the limit (speed=5000) it sometimes runs completely empty. This does not happen below the limit (speed=4000).

KevinOConnor commented:
On Mon, Jul 02, 2018 at 02:46:13PM -0700, Harald wrote:
> To me, one effect remains unexplained...
> Looking at https://github.com/KevinOConnor/klipper/issues/60#issuecomment-392387900, with accel=10000
> * a speed of 5000 creates continuous buffer_time peaks ("saw teeth") and a pause before each peak (where the buffer seems to be loaded)
> * with a speed of 4000 there are still peaks (a little bit lower), but there is no pause before a peak. Also, buffer_time never goes down to 0.000 (which is coincident with pausing).

In addition to the full flush refill goal of at least 2 seconds, after
every normal partial flush of the look-ahead queue, it tries to add
another .250 seconds of moves before rechecking the queue.  The "small
saw-tooth" is likely do to this.

The print pauses are likely due to OctoPrint (or similar) not being
able to fill commands fast enough.  Because the acknowledgment of
moves becomes saw-tooth like (no request for moves for several
seconds, and then a barrage of "ok" signals to request lots of moves
nearly instantaneously), it puts more load on the sender.  If the
sender can't get that block of moves to Klipper within a second or so,
then Klipper is forced to go into full flush mode (which can result in
a print pause and a big saw-tooth).  Thus, the more pronounced the
"small saw tooth" event is, the more likely the additional load will
result in a print pause.

In any case, it seems to be related to the unrealistic speeds.

-Kevin

hg42 commented:
I use Printrun/pronterface on linux as a host at 250000 baud.
I don't think there is a faster and more reliable method except sdcard printing.
However, I agree the baud rate is a speed limitation.

In the 5000mm/s case the first large saw-tooth is also preceded by a pause.
So I think this saw-tooth cannot be the reason for the pause. Instead it is a result of the pause phase.

However, you say the small saw-tooth would generate a lot of "ok" signals causing the pause and the large saw-tooth.
Ok...
But why doesn't the large saw-tooth do the same? And why doesn't it create much more "ok" signals?

hg42 commented:
being "offline" for a while, I am now on par with your code base again (including tests with my printer).
So I can continue here.

I have read the thread again and I think, I should eventually change the patch to what I suggested before:
> instead of `mainsync = section name` we could use a `mainsync = True` to assign it to one of the mcu sections.
> This would be easier for the user, because it removes the problem to exactly write the mcu section name (count of spaces).

In case of a missing flag, I would use `[mcu]` like before.

The only problem may be when the user marks two or more sections with `mainsync = True`.

What do you think?

hg42 commented:
I rebased the PR.

If creating an alternate implementation I would push it to a new PR.
However I think, the possibiity that a user can mark more than one sections as mainsync strongly suggests it would be the wrong way.

The problem I wanted to circumvent with that alternate implementation should probably be solved on it's own (or declared as a feature). I think, section names should be normalized when reading them (or different spaces should be allowed for separate names, which would be weird).

KevinOConnor commented:
I'm going to close this PR as it would need some work before merging and it doesn't look like it has enough priority to get that work done in the near future.

-Kevin
